app/code/Magento/CatalogUrlRewrite/Test/Unit/Model/Storage/DynamicStorageTest.php

```diff
@@ -0,0 +1,327 @@
+
+    /**
+     * @var DataObjectHelper|MockObject
+     */
+    private $dataObjectHelper;
+
+    /**
+     * @var AdapterInterface|MockObject
+     */
+    private $connection;
+
+    /**
+     * @var Select|MockObject
+     */
+    private $select;
+
+    /**
+     * @var ResourceConnection|MockObject
+     */
+    private $resourceConnection;
+
+    /**
+     * @var ScopeConfigInterface|MockObject
+     */
+    private $scopeConfig;
+
+    /**
+     * @var Product|MockObject
+     */
+    private $productResource;
+
+    /**
+     * @var ProductFactory|MockObject
+     */
+    private $productFactory;
+
+    /**
+     * @inheritdoc
+     */
+    protected function setUp(): void
+    {
+        $this->urlRewriteFactory = $this->getMockBuilder(UrlRewriteFactory::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $this->dataObjectHelper = $this->getMockBuilder(DataObjectHelper::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $this->connection = $this->getMockBuilder(AdapterInterface::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $this->select = $this->getMockBuilder(Select::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $this->resourceConnection = $this->getMockBuilder(ResourceConnection::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $this->connection
+            ->method('select')
+            ->willReturn($this->select);
+
+        $this->resourceConnection
+            ->method('getConnection')
+            ->willReturn($this->connection);
+
+        $this->scopeConfig = $this->getMockBuilder(ScopeConfigInterface::class)
+            ->getMock();
+
+        $this->productResource = $this->getMockBuilder(Product::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $this->productFactory = $this->getMockBuilder(ProductFactory::class)
+            ->disableOriginalConstructor()
+            ->getMock();
+
+        $this->productFactory
+            ->method('create')
+            ->willReturn($this->productResource);
+
+        $this->object = new DynamicStorage(
+            $this->urlRewriteFactory,
+            $this->dataObjectHelper,
+            $this->resourceConnection,
+            $this->scopeConfig,
+            $this->productFactory
+        );
+    }
+
+    /**
+     * @param $data
+     * @param $productFromDb
+     * @param $categorySuffix
+     * @param $categoryFromDb
+     * @param $canBeShownInCategory
+     * @param $expectedProductRewrite
+     * @throws \ReflectionException
+     * @dataProvider findProductRewriteByRequestPathDataProvider
+     */
+    public function testFindProductRewriteByRequestPath(
+        $data,
+        $productFromDb,
+        $categorySuffix,
+        $categoryFromDb,
+        $canBeShownInCategory,
+        $expectedProductRewrite
+    ) {
```

Comment:
Let's define object types 
```suggestion
    public function testFindProductRewriteByRequestPath(
        array $data,
        $productFromDb,
        string? $categorySuffix,
        $categoryFromDb,
        bool $canBeShownInCategory,
        array? $expectedProductRewrite
    ): void {
```